services/src/main/java/com/broadleafcommerce/bulkoperations/service/handler/CatalogBulkOperationHandler.java

```diff
@@ -0,0 +1,171 @@
+ *
+ * Licensed under the Broadleaf End User License Agreement (EULA), Version 1.1 (the
+ * "Commercial License" located at http://license.broadleafcommerce.org/commercial_license-1.1.txt).
+ *
+ * Alternatively, the Commercial License may be replaced with a mutually agreed upon license (the
+ * "Custom License") between you and Broadleaf Commerce. You may not use this file except in
+ * compliance with the applicable license.
+ *
+ * NOTICE: All information contained herein is, and remains the property of Broadleaf Commerce, LLC
+ * The intellectual and technical concepts contained herein are proprietary to Broadleaf Commerce,
+ * LLC and may be covered by U.S. and Foreign Patents, patents in process, and are protected by
+ * trade secret or copyright law. Dissemination of this information or reproduction of this material
+ * is strictly forbidden unless prior written permission is obtained from Broadleaf Commerce, LLC.
+ */
+package com.broadleafcommerce.bulkoperations.service.handler;
+
+import static com.broadleafcommerce.bulkoperations.service.environment.RouteConstants.Persistence.BULK_OPS_ROUTE_KEY;
+import static com.broadleafcommerce.common.messaging.service.DefaultMessageLockService.MESSAGE_IDEMPOTENCY_KEY;
+
+import org.apache.commons.codec.digest.DigestUtils;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.integration.support.MessageBuilder;
+import org.springframework.lang.Nullable;
+import org.springframework.messaging.Message;
+
+import com.broadleafcommerce.bulk.v2.domain.BulkOperationRequest;
+import com.broadleafcommerce.bulk.v2.domain.BulkOperationResponse;
+import com.broadleafcommerce.bulk.v2.domain.SupportedBulkOp;
+import com.broadleafcommerce.bulk.v2.messaging.BulkOpsInitializeItemsRequest;
+import com.broadleafcommerce.bulk.v2.messaging.BulkOpsInitializeItemsRequestProducer;
+import com.broadleafcommerce.bulk.v2.messaging.sandbox.CreateSandboxRequest;
+import com.broadleafcommerce.bulk.v2.messaging.sandbox.CreateSandboxRequestProducer;
+import com.broadleafcommerce.bulkoperations.domain.CatalogItem;
+import com.broadleafcommerce.bulkoperations.service.environment.BulkOperationsProviderProperties;
+import com.broadleafcommerce.bulkoperations.service.provider.CatalogProvider;
+import com.broadleafcommerce.common.extension.TypeFactory;
+import com.broadleafcommerce.common.messaging.notification.DetachedDurableMessageSender;
+import com.broadleafcommerce.data.tracking.core.context.ContextInfo;
+
+import java.security.SecureRandom;
+import java.util.List;
+import java.util.Random;
+
+import io.azam.ulidj.ULID;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.Setter;
+
+public class CatalogBulkOperationHandler implements BulkOperationHandler {
+
+    @Getter(AccessLevel.PROTECTED)
+    private final CatalogProvider<? extends CatalogItem> catalogProvider;
```

Comment:
Might be worth doing it this way if we expect to actually be reading CatalogItems so someone doesn't have to cast when extending.
```suggestion
public class CatalogBulkOperationHandler<CI extends CatalogItem> implements BulkOperationHandler<CI> {

    @Getter(AccessLevel.PROTECTED)
    private final CatalogProvider<CI> catalogProvider;
```

Reply:
I expect as this service grows that the BulkOperationHandler will be handling other types than CatalogItem. Catalog item is only needed on the provider call in the initialize items listener to get the entity context ids, not on this handler.